### PR TITLE
Combobox: Improve async UX

### PR DESCRIPTION
**Pull request by joshhunt**
This PR refactors a bunch of state management to improve the UX of async options loading. The expected user behaviour is:
 - reopening the menu after selecting a value should clear the text input so a value can be searched for easily
 - reopening the menu after selecting a value should show all options (options for empty input)
 - exiting the menu without selecting an item should restore the selected value to the text input.
 - onChange should not be triggered for any of the above unless a menu option is actually selected

This is achieved by no longer modifying downshift state in `on[Foo]Change` handlers, and instead intercepting state changes and making our own changes to it along the way using the state reducer prop.

A future PR will add a better loading state for an empty menu.

Part of https://github.com/grafana/grafana/issues/95936

### CODE REVIEW

**Comment by joshhunt** [comment]
@JoaoSilvaGrafana Can you confirm you're testing commit `5f962a5757c` 'clean up metriccombobox test'? I can't repro what you're talking about

https://github.com/user-attachments/assets/108a01c0-ae76-4fea-b731-92356f1d589b

**Comment by JoaoSilvaGrafana** [comment]
> @JoaoSilvaGrafana Can you confirm you're testing commit `5f962a5757c` `clean up metriccombobox test`? I can't repro what you're talking about
> 
>  8686_2024-11-08-16-57_firefox.mp4

Nope I didn't realise the git pull failed 😇 pulled it now and it's working! I also think the experience of using it is much better than before.

File: packages/grafana-ui/src/components/Combobox/Combobox.test.tsx [hunk 1]
```diff
@@ -294,5 +337,57 @@ describe('Combobox', () => {
 
       expect(emptyMessage).toBeInTheDocument();
     });
+
+    describe('with a value already selected', () => {
+      const selectedValue = { value: '1', label: 'Option 1' };
+
+      it('shows an empty text input when opening the menu', async () => {
```
Comment (tskarhed) [inline packages/grafana-ui/src/components/Combobox/Combobox.test.tsx:344]: In this case, do we want to set the placeholder if we already have a selected value?

**Review by tskarhed — COMMENTED** [review]
Not sure if this is directly related to these changes, but the "no labels" case shows labels. 
image

File: packages/grafana-ui/src/components/Combobox/Combobox.test.tsx (hunk 1 above)
Comment (joshhunt) [reply packages/grafana-ui/src/components/Combobox/Combobox.test.tsx:344]: I was in mixed minds about this, but we should match the old behaviour 👍

**Comment by joshhunt** [comment]
@tskarhed 

> but the "no labels" case shows labels

I don't believe it is - it's just that the values in that example have "label-like" values

```ts
const loadOptionsOnlyValues = useCallback((inputValue: string) => {
  return fakeSearchAPI(`http://example.com/search?query=${inputValue}`).then((options) =>
    options.map((opt) => ({ value: opt.label! }))
  );
}, []);
```

**Review by tskarhed — APPROVED** [review]

**Pull request merged**